File: algorithms/3_hard/M/00076_minimum_window_substring/sliding_window.cpp
typedef ptrdiff_t isize;

class Solution {
    static inline bool is_ok(const span<const isize> freqs) {
        for (char ch = 'A'; ch <= 'z'; ch += 1) {
            if (freqs[ch] < 0) {
                return false;
            }
        }
        return true;
    }

public:
    static string minWindow(const string& s, const string& t) {
        const isize s_len = std::ssize(s);
        if (s_len < std::ssize(t)) { return string(); }
        isize freqs[256] = { 0 };

        for (const auto ch : t) {
            freqs[ch] -= 1;
        }

        isize lo = 0;
        auto res = span<const char>(&s[0], 0);

        for (isize hi = 0; hi < s_len; hi += 1) {
            freqs[s[hi]] += 1;

            while (lo <= hi && freqs[s[lo]] > 0) {
                freqs[s[lo]] -= 1;
                lo += 1;
            }

            if ((res.empty() || std::ssize(res) > ((hi + 1) - lo)) && is_ok(freqs)) {
                res = span<const char>(&s[lo], (hi + 1) - lo);
            }
        }

        return string(res.cbegin(), res.cend());
    }
};

File: algorithms/3_hard/M/00076_minimum_window_substring/sliding_window.rs
impl Solution {
    #[inline(always)]
    fn is_ok(freqs: &[isize]) -> bool {
        for byte in b'A'..=b'z' {
            if freqs[byte as usize] < 0 {
                return false;
            }
        }
        true
    }

    pub fn min_window(s: String, t: String) -> String {
        let s_bytes = s.as_bytes();
        let s_len = s_bytes.len();
        if s_len < t.len() { return String::default(); }
        let mut freqs: [isize; 256] = [0; 256];

        for &byte in t.as_bytes() {
            freqs[byte as usize] -= 1;
        }

        let mut lo = 0;
        let mut res = &s_bytes[0..0];

        for hi in 0..s_len {
            freqs[s_bytes[hi] as usize] += 1;

            while lo <= hi && freqs[s_bytes[lo] as usize] > 0 {
                freqs[s_bytes[lo] as usize] -= 1;
                lo += 1;
            }

            if (res.is_empty() || res.len() > ((hi + 1) - lo)) && Self::is_ok(&freqs) {
                res = &s_bytes[lo..=hi];
            }
        }

        String::from_utf8(res.to_vec()).unwrap()
    }
}
